bcc/table.go

```diff
@@ -181,56 +185,134 @@ func (table *Table) DeleteAll() error {
 	return nil
 }
 
-// Iter returns a receiver channel to iterate over all table entries.
-func (table *Table) Iter() <-chan Entry {
+// TableIterator contains the current position for iteration over a *bcc.Table and provides methods for iteration.
+type TableIterator struct {
+	table *Table
+	fd    C.int
+
+	nextKey []byte
+	err     error
+
+	key     []byte
+	leaf    []byte
+	keyStr  []byte // lazily allocated
+	leafStr []byte // lazily allocated
+}
+
+// Iter returns an iterator to list all table entries available as raw bytes.
+func (table *Table) Iter() *TableIterator {
 	mod := table.module.p
-	ch := make(chan Entry, 128)
-	go func() {
-		defer close(ch)
-		fd := C.bpf_table_fd_id(mod, table.id)
-		key_size := C.bpf_table_key_size_id(mod, table.id)
-		leaf_size := C.bpf_table_leaf_size_id(mod, table.id)
-		key := make([]byte, key_size)
-		leaf := make([]byte, leaf_size)
-		keyP := unsafe.Pointer(&key[0])
-		leafP := unsafe.Pointer(&leaf[0])
-		alternateKeys := []byte{0xff, 0x55}
-		res := C.bpf_lookup_elem(fd, keyP, leafP)
-		// make sure the start iterator is an invalid key
-		for i := 0; i <= len(alternateKeys); i++ {
-			if res < 0 {
-				break
-			}
-			for j := range key {
-				key[j] = alternateKeys[i]
-			}
-			res = C.bpf_lookup_elem(fd, keyP, leafP)
-		}
-		if res == 0 {
-			return
-		}
-		keyStr := make([]byte, key_size*8)
-		leafStr := make([]byte, leaf_size*8)
-		keyStrP := (*C.char)(unsafe.Pointer(&keyStr[0]))
-		leafStrP := (*C.char)(unsafe.Pointer(&leafStr[0]))
-		for res = C.bpf_get_next_key(fd, keyP, keyP); res == 0; res = C.bpf_get_next_key(fd, keyP, keyP) {
-			r := C.bpf_lookup_elem(fd, keyP, leafP)
-			if r != 0 {
-				continue
-			}
-			r = C.bpf_table_key_snprintf(mod, table.id, keyStrP, C.size_t(len(keyStr)), keyP)
-			if r != 0 {
-				break
-			}
-			r = C.bpf_table_leaf_snprintf(mod, table.id, leafStrP, C.size_t(len(leafStr)), leafP)
-			if r != 0 {
-				break
-			}
-			ch <- Entry{
-				Key:   string(keyStr[:bytes.IndexByte(keyStr, 0)]),
-				Value: string(leafStr[:bytes.IndexByte(leafStr, 0)]),
-			}
+	fd := C.bpf_table_fd_id(mod, table.id)
+	keySize := C.bpf_table_key_size_id(mod, table.id)
+	leafSize := C.bpf_table_leaf_size_id(mod, table.id)
+	key := make([]byte, keySize)
+	leaf := make([]byte, leafSize)
+
+	nextKey := make([]byte, keySize)
+	keyP := unsafe.Pointer(&nextKey[0])
+
+	res := C.bpf_get_first_key(fd, keyP, keySize)
```

Comment:
I was wondering if this also could be done in `Next()`? Something like

```
if it.key == nil {
   // get first key
   return true
}
// get next key
```

Then `errNoIterationStartingPoint` wouldn't be needed and the implementation be closer to Go's `bufio.Scanner`.